feat(perf): implement container-level resource collection via cgroups

jra3 opened the pull request:
## Summary

This PR implements container-level resource monitoring via cgroup collectors, providing CPU and memory metrics for all running containers without requiring runtime API access.

## Key Features

- **Automatic cgroup version detection** (v1 and v2)
- **Multi-runtime support** (Docker, containerd, CRI-O, Podman)
- **Zero configuration** - discovers containers via filesystem
- **Graceful degradation** - handles missing files/permissions
- **Per-container metrics** - CPU usage, throttling, memory usage, limits

## Implementation Details

### Architecture
- **pkg/containers**: New package for centralized container discovery
- **Filesystem-based approach**: Direct cgroup reads for reliability
- **Per-container cgroup version tracking**: Support mixed environments

### Collectors
- **CgroupCPUCollector**: CPU usage, throttling, and pressure metrics
- **CgroupMemoryCollector**: Memory usage, limits, OOM detection, and pressure

## Changes Since Last Review

### ✅ Addressed Review Feedback
- ✨ **Documentation moved to wiki** with new structure under `Cgroup/` subdirectory
- 📝 **Converted all code examples to pseudocode** in documentation
- 📦 **Created `pkg/containers` package** for container discovery (moved from `pkg/performance/collectors`)
- 🔧 **Updated mount configuration** to use `/sys` instead of `/sys/fs/cgroup`
- 📚 **Added platform support clarification** (Linux-only)

### 🚀 Additional Improvements
- 📁 **Moved test script** from `test/` to `scripts/` directory
- 🎯 **Added kubectl context safety** with `--context` flag in test script
- 📊 **Added memory.peak support** for kernel 5.19+ (cgroup v2 equivalent of v1's max_usage_in_bytes)
- 🐛 **Fixed lint issues** (removed unused constant, handled filepath.Walk errors)

## Testing

```bash
# Run cgroup collector tests in KIND cluster
./scripts/test-cgroup-collectors.sh all

# Verify in production-like environment
make docker-build
make deploy
```

## Documentation

Full documentation available in the wiki:
- [Cgroup Collectors Overview](../../wiki/Cgroup)
- [Compatibility Guide](../../wiki/Cgroup/Compatibility)
- [CPU Collector](../../wiki/Cgroup/CPU-Collector)
- [Memory Collector](../../wiki/Cgroup/Memory-Collector)
- [Testing Guide](../../wiki/Cgroup/Testing)

## Checklist

- [x] Code follows project style guidelines
- [x] Tests pass locally
- [x] Documentation updated in wiki
- [x] Lint issues resolved
- [x] Review feedback addressed

haq204 commented on test/test-cgroup-collectors.sh:
```diff
@@ -0,0 +1,306 @@
+#!/usr/bin/env bash
```
Can we move this to `scripts/`?

jra3 commented:
### Resolved Review Feedback

The following review comments have been addressed in commit d41d709:

✅ **CLAUDE.md import syntax** - Updated to reference wiki documentation instead of local docs
✅ **Mount entire /sys** - Removed separate cgroup mount in agent.yaml, now using full /sys mount  
✅ **Documentation reorganization** - All cgroup docs moved to the project wiki with:
  - Simplified content without implementation code  
  - Better organization in Cgroup/ subdirectory
  - Removed overly detailed sections per feedback

The documentation feedback has been fully addressed. Working on remaining code-level suggestions next.

jra3 commented:
### Progress Update

✅ **Additional feedback addressed in commits d41d709 and 6b014ca:**

**Documentation & Configuration:**
- Removed cgroup docs from repo (moved to wiki)
- Updated agent.yaml to mount full /sys instead of separate cgroup mount
- Updated CLAUDE.md references

**Code improvements:**
- Removed unused MetricTypeProcessMemoryGrowth 
- Simplified cgroup path configuration to derive from HostSysPath
- Removed HostCgroupPath field throughout codebase
- Updated all tests to use new configuration

### Remaining items to address:
- Move container discovery to separate package
- Rename ContainerInfo to Container
- Make discovery structs private
- Address multi-runtime cgroup version handling

Working on these next...

jra3 commented:
## ✅ PR Update: All Feedback Addressed

Thank you for the thorough review! I've addressed all the feedback points. Here's a comprehensive summary:

### 🔧 Code Improvements

**Container Discovery Package**
- ✅ Created separate `pkg/containers` package as suggested
- ✅ Renamed `ContainerPath` to `Container` 
- ✅ Added per-container `CgroupVersion` field for mixed-runtime environments
- ✅ Made constants private (`minContainerIDLength`) following Go conventions

**Runtime Support**
- ✅ Improved CRI runtime detection (`cri-containerd`, `cri-o`)
- ✅ Added full Podman support (both systemd `libpod-<id>.scope` and non-systemd `/podman/<id>`)
- ✅ Now supports all major runtimes: Docker, containerd, CRI-O, and Podman

**Configuration**
- ✅ Updated to mount entire `/sys` instead of just `/sys/fs/cgroup`
- ✅ Removed `HostCgroupPath` field - now derives from `HostSysPath`
- ✅ Removed unused `MetricTypeProcessMemoryGrowth`

### 📚 Documentation Updates (Wiki)

All documentation has been moved to the wiki under `Cgroup/` subdirectory:
- [Overview](https://github.com/antimetal/system-agent/wiki/Cgroup-Overview)
- [Compatibility](https://github.com/antimetal/system-agent/wiki/Cgroup-Compatibility) 
- [CPU Collector](https://github.com/antimetal/system-agent/wiki/Cgroup-CPU-Collector)
- [Memory Collector](https://github.com/antimetal/system-agent/wiki/Cgroup-Memory-Collector)
- [Testing Guide](https://github.com/antimetal/system-agent/wiki/Cgroup-Testing)

**Documentation improvements:**
- ✅ Replaced all Go code with pseudocode/plain English
- ✅ Removed bash command examples per feedback
- ✅ Added platform support section (Linux-only, acknowledging production reality)
- ✅ Documented cgroupfs + systemd driver support (covers 90%+ of use cases)
- ✅ Added known limitations section (path stability, mount assumptions)
- ✅ Removed false claims about parallel scanning

### ✅ Testing
- All tests passing
- Fixed container ID extraction for Kubernetes pods
- Added test cases for Podman formats
- Updated test configurations for new structure

### 📝 Design Decisions Acknowledged

**Pragmatic choices we've made:**
- Sequential scanning is sufficient (no need for parallel complexity)
- Supporting cgroupfs + systemd covers virtually all production cases
- Linux-only is appropriate (production containers run on Linux)
- Filesystem approach over runtime APIs (simpler, more reliable)

**Known limitations (documented):**
- Cgroup paths may change across runtime versions
- Container IDs might theoretically be internal mappings
- Standard mount point assumed (customizable via HOST_SYS)

All critical feedback has been addressed. The PR is ready for final review!

jra3 commented:
## 🧪 KIND-Based Testing & Validation Guide

This PR includes a complete testing infrastructure for validating container cgroup collection in KIND clusters. Here's how to use it:

### Quick Start - Complete Test Environment

```bash
# 1. Create KIND cluster with test workloads
make cluster

# 2. Build and load agent
make build-and-load-image  

# 3. Deploy agent with cgroup collection enabled
make deploy
```

This gives you a complete environment with:
- ✅ **Antimetal agent** with cgroup filesystem access
- ✅ **Memory leak simulation** - gradually increasing memory usage
- ✅ **CPU stress testing** - generates throttling scenarios
- ✅ **Bursty workload patterns** - alternates between idle and busy
- ✅ **Multi-container deployment** - 5 nginx replicas for discovery testing

### Monitoring Test Workloads

**View active test containers:**
```bash
kubectl get pods -n antimetal-system -l test-suite=cgroup-collectors -o wide
```

**Monitor resource pressure (shows restarts from hitting limits):**
```bash
# Basic status with restart counts
kubectl get pods -n antimetal-system -l test-suite=cgroup-collectors

# Watch live updates
kubectl get pods -n antimetal-system -l test-suite=cgroup-collectors -w
```

**Check workload behavior:**
```bash
# Memory leak progression
kubectl logs -n antimetal-system memory-leak-test --tail=10

# CPU burst cycles  
kubectl logs -n antimetal-system bursty-workload -f

# Resource configuration
kubectl describe pods -n antimetal-system -l test-suite=cgroup-collectors
```

### Validating Cgroup Collection

**Verify agent has cgroup access:**
```bash
kubectl describe pod -n antimetal-system -l app.kubernetes.io/name=agent | grep -A 5 cgroup
```

**Test comprehensive functionality:**
```bash
# Run full test suite (deploys agent + validates collection)
./scripts/test-cgroup-collectors.sh all

# Monitor collector activity
./scripts/test-cgroup-collectors.sh monitor

# Clean up when done
./scripts/test-cgroup-collectors.sh cleanup
```

### Expected Test Scenarios

The test workloads create realistic container resource scenarios:

1. **🔄 CPU Throttling**: `cpu-stress-test` runs 2 CPU workers with 500m limit → triggers cgroup CPU throttling
2. **📈 Memory Pressure**: `memory-leak-test` gradually increases allocation → tests memory cgroup limits  
3. **💥 Resource Limits**: Workloads restart when hitting limits → validates OOM detection
4. **🚀 Container Discovery**: 5 nginx replicas → tests multi-container enumeration
5. **⚡ Dynamic Patterns**: Bursty workload → tests collection of varying resource usage

### Architecture Notes

- **Test workloads** are now in `config/test/` as proper Kustomize components
- **Agent configuration** includes `/sys/fs/cgroup` mount at `/host/cgroup`  
- **Default deployment** includes both agent and test workloads via `../test` in kustomization
- **Targeted patching** ensures metrics patches only apply to agent, not test deployments

This provides a complete, reproducible environment for validating container-level cgroup collection without requiring external infrastructure or manual container setup.

jra3 commented:
## 🧪 KIND-Based Testing & Validation Guide (Updated)

This PR includes a complete testing infrastructure for validating container cgroup collection in KIND clusters. Here's the simplified workflow:

### Complete Test Environment Setup

```bash
# 1. Create KIND cluster with test workloads  
make cluster

# 2. Build and load agent
make build-and-load-image  

# 3. Deploy agent with cgroup collection enabled
make deploy
```

This gives you a complete environment with:
- ✅ **Antimetal agent** with cgroup filesystem access (`/sys/fs/cgroup` → `/host/cgroup`)
- ✅ **Memory leak simulation** - gradually increasing memory usage
- ✅ **CPU stress testing** - generates throttling scenarios  
- ✅ **Bursty workload patterns** - alternates between idle and busy
- ✅ **Multi-container deployment** - 5 nginx replicas for discovery testing

### Monitoring Test Workloads

**View test workloads:**
```bash
kubectl get pods -n antimetal-system -l test-suite=cgroup-collectors -o wide
```

**Monitor workload behavior:**
```bash
# Memory leak progression
kubectl logs -n antimetal-system memory-leak-test --tail=10

# CPU burst cycles  
kubectl logs -n antimetal-system bursty-workload -f

# Watch for restarts (resource pressure)
kubectl get pods -n antimetal-system -l test-suite=cgroup-collectors -w
```

**Check resource limits and pressure:**
```bash
kubectl describe pods -n antimetal-system -l test-suite=cgroup-collectors
```

### Validating Cgroup Collection

**Verify agent configuration:**
```bash
kubectl describe pod -n antimetal-system -l app.kubernetes.io/name=agent | grep -A 5 cgroup
```

**Monitor agent logs:**
```bash
kubectl logs -n antimetal-system deployment/agent --tail=100 -f
```

**Check collector activity:**
```bash
kubectl logs -n antimetal-system deployment/agent --tail=1000 | grep -i "performance\|collector\|cgroup"
```

### Expected Test Scenarios

The test workloads create realistic container resource scenarios:

1. **🔄 CPU Throttling**: `cpu-stress-test` runs 2 CPU workers with 500m limit → triggers cgroup CPU throttling
2. **📈 Memory Pressure**: `memory-leak-test` gradually increases allocation → tests memory cgroup limits  
3. **💥 Resource Limits**: Workloads restart when hitting limits → validates OOM detection
4. **🚀 Container Discovery**: 5 nginx replicas → tests multi-container enumeration
5. **⚡ Dynamic Patterns**: Bursty workload → tests collection of varying resource usage

### Clean Up

```bash
make destroy-cluster  # Remove entire cluster
```

### Architecture Notes

- **Test workloads** are in `config/test/` as Kustomize components, automatically included in deployment
- **Agent configuration** includes `/sys/fs/cgroup` mount at `/host/cgroup` with `HOST_CGROUP` env var
- **Targeted patching** ensures agent metrics patches don't interfere with test deployments
- **Standard kubectl** provides all necessary monitoring and validation capabilities

This provides a complete, reproducible environment for validating container-level cgroup collection using standard Kubernetes tooling.
